// xml-dom/main.js
module.exports = {
  DOM: function(xmlObj) {
    var self = this;
    self.document = {
      childNodes: typeof xmlObj == "object" ? xmlObj : module.exports.parse(xmlObj)
    };
    var domMethods = {
      getElementsByTagName: function(tagName) {
        return troughChilds(this, function(child) {
          if (child.tagName == tagName) return true;
          return false;
        });
      },
      getElementsByAttribute: function(name, value) {
        return troughChilds(this, function(child) {
          if (child.attributes && child.attributes[name]) {
            if (child.attributes[name].match(value)) {
              return true;
            }
          }
          return false;
        });
      },
      removeChild: function(childNode) {
        var self = this;
        troughChilds(self, function(child) {
          if (child === childNode) {
            var index = self.childNodes.indexOfKey(child.index, "index");
            delete self.childNodes[index];
            self.childNodes = (function() {
              var newArray = [];
              self.childNodes.forEach(function(item) {
                if (item) newArray.push(item);
              });
              return newArray;
            })();
            return true;
          }
          return false;
        });
        return true;
      },
      appendChild: function(childNode) {
        this.childNodes.push(childNode);
        appendPrototypes(childNode, this);
      },
      insertBefore: function(childNode, nodeAfter) {
        var childs = [];
        for (var i = 0; i < this.childNodes.length; i++) {
          if (this.childNodes[i] === nodeAfter) {
            childs.push(childNode);
            appendPrototypes(childNode, this);
          }
          childs.push(this.childNodes[i]);
        }
        this.childNodes = childs;
      },
      getElementsByCheckFunction: function(check) {
        return troughChilds(this, check);
      }
    }
    self.document = appendPrototypes(self.document, self);
    troughChilds(self.document, function(child) {
      if (child) {
        appendPrototypes(child, self.document);
      }
    });
    function appendPrototypes(node, parent) {
      Object.keys(domMethods).forEach(function(methodName) {
        node[methodName] = domMethods[methodName];
      });
      node["parentNode"] = parent;
      return node;
    }
    function troughChilds(parent, checkFunction) {
      var children = parent.childNodes;
      var childNodes = [];
      children.forEach(function(child) {
        checkChild(child, parent);
      });
      function checkChild(child, parent) {
        if (checkFunction(child) == true) {
          child = appendPrototypes(child, parent);
          childNodes.push(child);
        }
        if (child.type === "element") {
          child.childNodes.forEach(function(childNode) {
            checkChild(childNode, child);
          });
        }
      }
      return childNodes;
    }
  }
}
Array.prototype.indexOfKey = function(value, key) {
  for (var i = 0; i < this.length; i++) {
    if (this[i][key] === value) {
      return i;
    }
  }
  return -1;
}
